lib/kdtree.cpp

```diff
@@ -301,6 +302,82 @@ KDTree::find_plane_and_classify(const TriangleIds& tris, const Box& box) const {
                            std::move(ltris), std::move(rtris));
 }
 
+void KDTree::flatten() {
+    static constexpr uint32_t INVALID_INDEX = 0xFFFFFFFF >> 2;
+
+    // auto print = [this] {
+    //     size_t i = 0;
+    //     for (auto it = nodes_.begin(); it != nodes_.end(); ++it, ++i) {
+    //         const auto& node = *it;
+    //         if (node.is_inner()) {
+    //             std::cerr << i << ": InnerNode("
+    //                       << static_cast<int>(node.split_axis()) << ", "
+    //                       << node.split_pos() << ", " << node.right() << ")"
+    //                       << std::endl;
+    //         } else {
+    //             if (node.is_empty()) {
+    //                 std::cerr << i << ": Leaf()" << std::endl;
+    //             } else if (node.is_sentinel()) {
+    //                 std::cerr << i << ": Leaf(" << node.first_triangle_id()
+    //                           << ")" << std::endl;
+    //             } else {
+    //                 std::cerr << i << ": Leaf(" << node.first_triangle_id()
+    //                           << ", " << node.second_triangle_id() << ")"
+    //                           << std::endl;
+    //             }
+    //         }
+    //     }
+    // };
+
+    // do DFS through nodes
+    std::stack<std::pair<Node*, uint32_t /*parent index*/>> stack;
+    stack.emplace(root_.get(), INVALID_INDEX);
+
+    while (!stack.empty()) {
+        const Node& node = *stack.top().first;
+        uint32_t parent_index = stack.top().second;
+        stack.pop();
+
+        uint32_t node_index = nodes_.size();
+        if (parent_index != INVALID_INDEX) {
+            // we are in the right node => update parent's reference
+            nodes_[parent_index].set_right(node_index);
+        }
+
+        if (node.is_inner()) {
+            // set right to invalid, since we don't know the index yet
+            nodes_.emplace_back(node.split_axis(), node.split_pos(),
+                                INVALID_INDEX);
+
+            // Push right first, since we visit left first.
+            stack.emplace(node.right(), node_index);
+            stack.emplace(node.left(), INVALID_INDEX);
+        } else {
+            bool is_odd = true;
+            TriangleId first_id;
+            const TriangleId* triangle_ids = node.triangle_ids();
+            for (size_t i = 0; i < node.num_tris(); ++i) {
+                if (is_odd) {
+                    first_id = triangle_ids[i];
+                } else {
+                    nodes_.emplace_back(first_id, triangle_ids[i]);
+                }
+                is_odd = !is_odd;
```

Comment:
What's happening here? Are we skipping triangles? It's not easy to grasp the layout of the flattened tree. It seems as if the leafs are binary trees.